### PR TITLE
Fixes oct24

### PR DESCRIPTION
**Pull request by CompiledAtBirth**
**Lampe and SBI backends only**

The main goal of this PR is the [adatpation](https://github.com/maho3/ltu-ili/issues/165) of the software to sbi versions greater than 0.22.

I also modified the names accordingly in the Lampe backend, so that when returning the _summaries_ from a 
[LampeRunner](https://github.com/maho3/ltu-ili/blob/main/ili/inference/runner_lampe.py) or [SBIRunner](https://github.com/maho3/ltu-ili/blob/main/ili/inference/runner_sbi.py), the dictionary keys are consistent.

I also took the liberty to add a "repeats" arguments [not to duplicate code](https://github.com/maho3/ltu-ili/issues/42)  for networks instantiation when training large ensembles.
This _repeats_ argument intervenes in [utils/ndes_pt](https://github.com/maho3/ltu-ili/blob/main/ili/utils/ndes_pt.py) in the direct python interface, and for configuration files this new argument is handled in the runner instantiations.

I also modified the examples accordingly (in the notebooks and examples folders), but have not exhaustively tested.

### CODE REVIEW

**Comment by CompiledAtBirth** [comment]
Tests and examples with sbi and lampe backends have been modified accordingly.

File: ili/embedding/fcn.py [hunk 1]
```diff
@@ -56,3 +52,6 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
             self.initalize_model(x.shape[-1])
 
         return self.mlp(x)
+
+    # def parameters(self, recurse = True):
```
Comment (maho3) [inline ili/embedding/fcn.py:56]: Delete?

File: ili/embedding/fcn.py [hunk 2]
```diff
@@ -12,24 +12,20 @@ class FCN(nn.Module):
...
         n_hidden (List[int]): number of hidden units per layer
         act_fn (str):  activation function to use
     """
 
     def __init__(
-        self, n_hidden: List[int], act_fn: str = "SiLU"
+        self,n_input, n_hidden: List[int], act_fn: str = "SiLU"
     ):
-        super().__init__()
+        #super().__init__()
```
Comment (maho3) [inline ili/embedding/fcn.py:23]: Delete?

File: ili/embedding/fcn.py [hunk 3]
```diff
@@ -12,24 +12,20 @@ class FCN(nn.Module):
...
+        super(FCN,self).__init__()
         self.act_fn = getattr(nn, act_fn)()
         self.n_layers = len(n_hidden)
         self.n_hidden = n_hidden
 
-    def initalize_model(self, n_input: int):
```
Comment (maho3) [inline ili/embedding/fcn.py:27]: Why was initialize_model removed? Perhaps because it no longer works with sbi?

This is quite useful for not needing to specify the n_input sometimes (especially in Lampe), and we should leave it if it doesn't conflict.

Perhaps you could leave it, and if an optional `n_input` is specified in init then init can call it?

File: ili/inference/runner_sbi.py [hunk 4]
```diff
@@ -9,18 +9,19 @@
 import pickle
 import torch
 import torch.nn as nn
+from torch import Tensor
 from pathlib import Path
 from typing import Dict, List, Callable, Optional, Union
 from torch.distributions import Distribution
 from sbi.inference import NeuralInference
-from sbi.utils.posterior_ensemble import NeuralPosteriorEnsemble
+#from sbi.utils.posterior_ensemble import NeuralPosteriorEnsemble #"NeuralPosteriorEnsemble was renamed EnsemblePosterior and moved to sbi.inference.posteriors.ensemble_posterior. sbi.utils.posterior_ensemble
```
Comment (maho3) [inline ili/inference/runner_sbi.py:17]: Delete

File: ili/inference/runner_sbi.py [hunk 5]
```diff
@@ -236,14 +258,14 @@ def _train_round(self, models: List[NeuralInference],
             summaries.append(model.summary)
 
         # ensemble all trained models, weighted by validation loss
-        val_logprob = torch.tensor(
-            [float(x["best_validation_log_prob"][-1]) for x in summaries]
-        ).to(self.device)
+        val_logprob = torch.tensor([float(x["best_validation_loss"][-1]) for x in summaries]).to(self.device)
```
Comment (maho3) [inline ili/inference/runner_sbi.py:261]: Shouldn't this be negative loss? to get the logprob?

File: ili/inference/runner_sbi.py [hunk 6]
```diff
@@ -236,14 +258,14 @@ def _train_round(self, models: List[NeuralInference],
...
+        
         # Exponentiate with numerical stability
         weights = torch.exp(val_logprob - val_logprob.max())
         weights /= weights.sum()
-
-        posterior_ensemble = NeuralPosteriorEnsemble(
+        #print(weights is Tensor) # sbi wrote this, need to raise issue/make PR
```
Comment (maho3) [inline ili/inference/runner_sbi.py:266]: Delete

File: ili/utils/ndes_pt.py [hunk 7]
```diff
@@ -66,20 +67,20 @@ def load_nde_sbi(
         # check for arguments
         if not (set(model_args.keys()) <= {'hidden_features', 'num_transforms'}):
             raise ValueError(f"Model {model} arguments mispecified.")
-
+    #Please use `from sbi.neural_nets import posterior_nn` in the future (not sbi.utils.posterior_nn)
```
Comment (maho3) [inline ili/utils/ndes_pt.py:70]: Delete?

File: ili/utils/ndes_pt.py [hunk 8]
```diff
@@ -250,6 +251,9 @@ def sample(
             for nde, N in zip(self.posteriors, per_model)
         ], dim=0)
         samples = samples[:num_samples]
+        # Jan. 2025: debugging purpose for PlotSinglePosterior that sometimes fails in the Lampe Backend
+        if num_samples != shape[0]:
+            print(num_samples)
```
Comment (maho3) [inline ili/utils/ndes_pt.py:256]: Delete?

File: ili/utils/samplers.py [hunk 9]
```diff
@@ -15,7 +15,9 @@
 try:
     import torch
     from sbi.inference.posteriors.base_posterior import NeuralPosterior
-    from sbi.utils.posterior_ensemble import NeuralPosteriorEnsemble
+    #from sbi.utils.posterior_ensemble import NeuralPosteriorEnsemble
```
Comment (maho3) [inline ili/utils/samplers.py:18]: Delete?

File: ili/validation/metrics.py [hunk 10]
```diff
@@ -19,7 +19,8 @@
 try:
     from sbi.inference.posteriors.base_posterior import NeuralPosterior
     from sbi.inference.posteriors import DirectPosterior
-    from sbi.utils.posterior_ensemble import NeuralPosteriorEnsemble
+    from sbi.inference.posteriors import EnsemblePosterior
+    #from sbi.utils.posterior_ensemble import NeuralPosteriorEnsemble
```
Comment (maho3) [inline ili/validation/metrics.py:23]: Delete?

File: ili/validation/metrics.py [hunk 11]
```diff
@@ -95,10 +96,11 @@ def _build_sampler(self, posterior: ModelClass) -> ABC:
                 'Pyro backend is only available for sbi posteriors')
 
         # check if DirectPosterior is available
+        print(type(posterior))
```
Comment (maho3) [inline ili/validation/metrics.py:99]: Delete?

File: ili/validation/runner.py [hunk 12]
```diff
@@ -12,11 +12,13 @@
 from ili.dataloaders import _BaseLoader
 from ili.validation.metrics import _BaseMetric
 from ili.utils import load_from_config, update
-
+#from sbi.utils.posterior_ensemble import NeuralPosteriorEnsemble #"NeuralPosteriorEnsemble was renamed EnsemblePosterior and moved to sbi.inference.posteriors.ensemble_posterior. sbi.utils.posterior_ensemble
```
Comment (maho3) [inline ili/validation/runner.py:15]: Delete?

File: ili/validation/runner.py [hunk 13]
```diff
@@ -12,11 +12,13 @@
...
+#from sbi.inference.posteriors import EnsemblePosterior
 try:
     from sbi.inference.posteriors.base_posterior import NeuralPosterior
     ModelClass = NeuralPosterior
-    from sbi.utils.posterior_ensemble import NeuralPosteriorEnsemble
+    #from sbi.utils.posterior_ensemble import NeuralPosteriorEnsemble
```
Comment (maho3) [inline ili/validation/runner.py:20]: Delete?

File: notebooks/sbi.ipynb
Comment (maho3) [inline notebooks/sbi.ipynb:1]: I need to check this in VSCode later on...

File: setup.cfg [hunk 14]
```diff
@@ -23,7 +23,8 @@ install_requires =
 
     # pytorch backend
     torch;python_version>='3.7'
-    sbi<=0.22.0;python_version>='3.7'
+    torch_geometric;python_version>='3.7'
```
Comment (maho3) [inline setup.cfg:26]: torch_geometric is not strictly necessary unless you want to do graph embedding networks. I think it also conflicts with some cuda environments, so best to leave it out

File: tests/test_sbi.py [hunk 15]
```diff
@@ -28,6 +28,10 @@
 device = 'cuda' if torch.cuda.is_available() else 'cpu'
 print('Device:', device)
 
+def test_dummy(monkeypatch):
+    monkeypatch.setattr(plt, 'show', lambda: None)
+    print("HELLO")
```
Comment (maho3) [inline tests/test_sbi.py:33]: Are we keeping this?

**Review by maho3 — CHANGES_REQUESTED** [review]
This looks really great Nicolas! I think its nearly ready to merge. Notes:

- I highlighted a few places where old comments and tests are leftover that should probably be cleaned up.
- I wonder if, for renamed `sbi` things (e.g. `NeuralPosteriorEnsemble` -> `EnsemblePosterior`) we should do conditional imports to allow for backwards compatibility. Something like
```python
try:
    from sbi.utils.posterior_ensemble import NeuralPosteriorEnsemble as EnsemblePosterior
except ImportError:
    from sbi.inference.posteriors import EnsemblePosterior
```
- I want to run a formatting linter on this, which I will do just before we want to merge it.

File: tests/test_sbi.py (hunk 15 above)
Comment (CompiledAtBirth) [reply tests/test_sbi.py:33]: No need

File: ili/embedding/fcn.py (hunk 3 above)
Comment (CompiledAtBirth) [reply ili/embedding/fcn.py:27]: I will test your idea to have init call initialize with optional argument.

Comment (CompiledAtBirth) [reply ili/embedding/fcn.py:27]: With the previous [fcn](https://github.com/maho3/ltu-ili/blob/main/ili/embedding/fcn.py) class, the generator object FCN.parameters does not return anything:

`embedding_net = FCN(**embedding_args)`
`for xdd in embedding_net.parameters():
    print(xdd, "\n")`
    
 Which throws an error when [sbi.utils.nn_utils.check_net_device](https://github.com/sbi-dev/sbi/blob/main/sbi/utils/nn_utils.py#L46) is called when instantiating `sbi.neural_nets`, because net.parameters() does not return anything. 
image

Comment (CompiledAtBirth) [reply ili/embedding/fcn.py:27]: I propose the solution of this PR, with n_input as an optional parameter set to 0 by default such that:
`n_input = n_hidden[0] if n_input == 0 else n_input`.

File: ili/inference/runner_sbi.py (hunk 5 above)
Comment (CompiledAtBirth) [reply ili/inference/runner_sbi.py]: Well spotted, I corrected this.

File: ili/embedding/fcn.py (hunk 3 above)
Comment (CompiledAtBirth) [reply ili/embedding/fcn.py:27]: Having a default n_input could throw errors if, for instance using ltu-ili in a notebook setting, one makes some changes to the input vector dimension before training. For now I would have n_input as an explicit instantiation argument. What do you think ?

Comment (maho3) [reply ili/embedding/fcn.py:27]: Ah thanks for digging into this. Hmm, if I remember correctly, it was their models that required an `initialize_model` function first, and they now seem to be doubling back on it.

I agree, it is better practice to have the model definition in place upon `__init__`. However, removing the automatic reshaping will break some of our legacy implementations (especially a problem for non-LtU people using this).

So, I propose the following: We add `n_input=None` to the `__init__` constructor. If it's not `None` then we internally call `initialize_model` with the specified `n_input`. If it is `None`, then we do not setup the network until `__call__`, but we add a hack where we have a dummy parameter which will appear when we call `net.parameters()` but doesn't get used in the inference. For example:
```python
def __init__():
    ...
    self._dummy = nn.Parameter(0, device=device)
```

It's quite hacky, but will solve our problem and allow for backwards compatibility

Comment (CompiledAtBirth) [reply ili/embedding/fcn.py:27]: Your hack works ! We just have to make it explicitly a tensor to have a 'detach' method

` self.dummy = nn.Parameter(torch.Tensor([0]),requires_grad = False) `

Comment (maho3) [reply ili/embedding/fcn.py:27]: Awesome ! I'm glad this works

**Review by maho3 — COMMENTED** [review]
This looks very good. I want to go through and remove old comments and fix formatting, but otherwise its good to merge.
